boefjes/boefjes/app.py

```diff
@@ -31,109 +28,192 @@ def __init__(
     ):
         self.item_handler = item_handler
         self.client_factory = client_factory
+        self.scheduler_client = client_factory()
         self.settings = settings
+        self.task_queue = mp.Queue(maxsize=self.settings.pool_size)
+
+        manager = mp.Manager()
+        self.handling_tasks = manager.dict()
+
+        self.worker_args = (self.task_queue, self.item_handler, self.client_factory, self.handling_tasks)
 
         logger.setLevel(log_level)
 
-    def run(self, queue: RuntimeManager.Queue) -> None:
-        logger.info("Creating worker pool for queue '%s'", queue.value)
-        pool_size = self.settings.pool_size
+    def run(self, queue_type: WorkerManager.Queue) -> None:
+        logger.info("Created worker pool for queue '%s'", queue_type.value)
+
+        workers = [mp.Process(target=_start_working, args=self.worker_args) for _ in range(self.settings.pool_size)]
 
-        with Pool(processes=pool_size) as pool:
+        for worker in workers:
+            worker.start()
+
+        while True:
             try:
-                # Workers pull tasks from the scheduler
-                pool.starmap(
-                    start_working,
-                    [(self.client_factory(), self.item_handler, self.settings, queue) for _ in range(pool_size)],
-                )
-            except Exception:  # noqa
-                logger.exception("An error occurred")
-
-            logger.info("Closing worker pool")
-
-
-def start_working(
-    scheduler_client: SchedulerClientInterface,
-    item_handler: Handler,
-    settings: Settings,
-    queue_to_handle: RuntimeManager.Queue,
-) -> None:
-    """
-    This function runs in parallel and polls the scheduler for queues and jobs.
-    Hence, it should catch most errors and give proper, granular feedback to the user.
-    """
-
-    logger.info("Started worker process [pid=%d]", os.getpid())
+                workers = self._check_workers(workers)
+                self._fill_queue(self.task_queue, queue_type)
+            except Exception as e:  # noqa
+                logger.exception("Unhandled Exception:")
+                logger.info("Continuing worker...")
+                continue
+            except BaseException as e:  # noqa
+                logger.exception("Exiting worker...")
+                for worker in workers:
+                    worker.terminate()
+                    worker.join()
+                    worker.close()
+
+                raise
+
+    def _fill_queue(self, task_queue: mp.Queue, queue_type: WorkerManager.Queue):
+        if task_queue.full():
+            time.sleep(self.settings.poll_interval)
+            return
```

Comment:
I don't know what the supposed purpose of this sleep is (adding comments to code like this would help). Putting something on the queue already blocks if the queue is full, so we don't need to check if the queue is full and sleep for that. If the poll interval is higher sleeping for the poll interval will also mean the throughput is lower if tasks are executed faster than the poll interval.

If it is supposed to run `_check_workers` to make sure that we don't block when all the workers have died, this won't work reliably with multiple organisations because we can still block when we put multiple tasks on the queue later. I think it would be better to call `task_queue.put` with a timeout (which can be large, because this isn't supposed to happen) and run `_check_workers` if there is a timeout and then try to put the task on the queue again.

Reply:
Good spot on the multiple organizations issue! I believe this is partly an artifact from a prior implementation and partly indeed to quickly spot broken workers. Checking the workers when the queue does not shrink for a while makes sense, but for large timeouts `t` we could theoretically block for `t` seconds in a worst case scenario.

Another possible solution would be to remove the `maxsize` and only fill the queue when `task_queue.qsize()` is below `n * #workers` for some `n`. Then we would check the workers every `self.settings.poll_interval` to quickly drop hanging workers and only lose some efficiency if all workers would finish `n` tasks in the interval (or, on average at least). @dekkers Thoughts?

I also realized that there is no graceful handling of dispatched tasks on the queue on shutdown.. (Which would be even more relevant if we increase the  queue size.) I Will have a look at that as well.

Reply:
I think not setting maxsize makes sense as that would prevent blocking when putting things on the queue.

If we want to sleep when all workers are busy I think it would better to introduce another setting for this as the length of poll interval is not relevant for this. For example if you have lots of small organisations you might set the poll interval higher to not hammer the scheduler, but that doesn't mean you want to sleep that long when all workers are busy. Maybe we should also sleep shorter intervals and check again if the queue is still full multiple times before checking the workers. E.g. sleep 1 sec, check queue, sleep 1 sec, check queue, ..., slept more than 10 times? check workers or something like that. That would prevent sleeping too long.

Reply:
A separate setting makes sense. Perhaps in the end we actually just want to be able to define a **rate limit** on the scheduler, a **heartbeat interval** on the workers and a **poll interval** on the queue, but that would perhaps require separate threads to keep things readable. Also, only checking the workers when the queue is full for 10 iterations would open up the possibility that we only detect a dead worker when the load is high, if I understand your suggestion correctly.

I'll create a separate setting and perhaps a feature request towards the scheduler to look into introducing an endpoint that would give exactly what these workers, at least in the common case, need: per `queue_type` at most `n` items that could immediately be dispatched to the multiprocessing queue ( E.g. `POST queues/pop?limit=3` vs. the current `POST queues/{queue_id}/pop`). That would reduce the calls to the scheduler significantly when you have a lot of organisations and also make this configuration value actually represent the real interval with which we ping the scheduler.

Reply:
I think the maxsize of the queue still needs to be removed to prevent potential blocking in calling `put`. The check in `_fill_queue` will then mean that the maximum number of tasks on the queue is pool_size+number of organisations.

Reply:
Ah indeed

Reply:
Done